File: agent-driven-architecture-main/examples/typescript/src/blocks/inbox/fold.ts
// ── blocks/inbox/fold — the block's ARM (12.4) ─────────────────────────────
// Two `never`-guarded matches, nested: one over the block's own verbs, one over
// its own sealed `DropReason`. Adding a third drop reason breaks the build here
// and in `project.ts`, and nowhere outside this folder.

import type { Signature } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";
import type { ArmOut } from "@adr/spine/pure/verb";
import { armOut } from "@adr/spine/pure/verb";
import type { DropReason, InboxResult } from "./contract";
import type { InboxSlice } from "./slice";
import { withConflated, withDuplicate, withFault } from "./slice";

export function inboxArm(
  slice: InboxSlice,
  r: InboxResult,
  now: Timestamp,
  _sig: Signature,
): ArmOut<InboxSlice> {
  switch (r.tool) {
    case "noteDrop":
      return armOut(applyDrop(slice, r.reason), [], []);
    case "noteFault":
      return armOut(withFault(slice, { at: now, source: r.source, fault: r.fault }), [], []);
    default: {
      const _never: never = r;
      return _never;
    }
  }
}

function applyDrop(slice: InboxSlice, reason: DropReason): InboxSlice {
  switch (reason.kind) {
    case "Conflated":
      return withConflated(slice, reason.source, reason.dropped);
    case "Duplicate":
      return withDuplicate(slice, reason.source);
    default: {
      const _never: never = reason;
      return _never;
    }
  }
}

File: agent-driven-architecture-main/examples/kotlin/block/inbox/src/main/kotlin/adr/blocks/inbox/Fold.kt
// ── blocks/inbox/fold — the block's ARM ────────────────────────────────────
// Exhaustive over the block's sealed sub-union, with NO else arm. The same three
// rules every arm follows (§7). Neither verb emits an effect: a dropped input is a
// fact to record, not an action to take — and RunStatus is untouchable from a block
// anyway (gate check C6), which is what keeps one shed input from degrading the
// whole session banner (12.4).

package adr.blocks.inbox

import adr.contract.InboxResult
import adr.spine.pure.ArmOut
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp

internal class InboxArm {

    fun arm(
        slice: InboxSlice,
        result: InboxResult,
        now: Timestamp,
        sig: Signature,
    ): ArmOut<InboxSlice> = when (result) {
        is InboxResult.NoteDrop ->
            if (result.dropped <= 0) {
                // RULE 1: the arm validates against reality before it decides. "Zero
                // dropped" is not a drop, and recording it would inflate the counter the
                // operator uses to decide whether the tier is overloaded.
                ArmOut(
                    slice = slice,
                    notices = listOf(Notice.Rejected(now, result.tool, "a drop count must be positive")),
                )
            } else {
                ArmOut(slice = slice.withDrop(result.source, result.reason, result.dropped))
            }

        is InboxResult.NoteFault ->
            ArmOut(slice = slice.withFault("${result.source.value}: ${result.fault}"))
    }
}
